src/shared/utilities/asyncCollection.ts

```diff
@@ -0,0 +1,190 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Coalesce } from './tsUtils'
+
+/**
+ * High-level abstraction over async generator functions of the form `async function*` {@link AsyncGenerator}
+ */
+export interface AsyncCollection<T> extends AsyncIterable<T> {
+    /** Flattens the collection 1-level deep */
+    flatten(): AsyncCollection<SafeUnboxIterable<T>>
+    /** Applies a mapping transform to the output generator */
+    map<U>(fn: (obj: T) => U): AsyncCollection<U>
+    /** Filters out results which will _not_ be passed on to further transformations. */
+    filter<U extends T>(predicate: AsyncPredicate<T, U>): AsyncCollection<U>
+    /** Uses only the first 'count' number of values returned by the generator. */
+    take(count: number): AsyncCollection<T>
+    /** Converts the collection to a Promise, resolving to an array of all values returned by the generator. */
+    promise(): Promise<T[]>
+    /** Converts the collection to a Map, using either a property of the item or a function to select keys. */
+    toMap<K extends StringProperty<T>, U extends string = never>(
+        selector: KeySelector<T, U> | K
+    ): Promise<Map<Coalesce<U, T[K]>, T>>
+    iterator(): AsyncIterator<T, T | void>
+}
+
+/**
+ * Converts an async generator function to an {@link AsyncCollection}
+ *
+ * Uses closures to capture generator functions after each transformation. Generator functions are not called
+ * until a 'final' operation is taken by either:
+ *  * Iterating over them using `for await (...)`
+ *  * Iterating over them using `.next()`
+ *  * Calling one of the conversion functions `toMap` or `promise`
+ *
+ * Collections are *immutable* in the sense that any transformation will not consume the underlying generator
+ * function. That is, any 'final' operation uses its own contextually bound generator function separate from
+ * any predecessor collections.
+ */
+export function toCollection<T>(generator: () => AsyncGenerator<T, T | undefined | void>): AsyncCollection<T> {
```

Comment:
+1 for avoiding verbosity, but it seems important to mention "async" or "lazy" in this function name. That helps a lot when reading the call sites.
```suggestion
export function toAsyncCollection<T>(generator: () => AsyncGenerator<T, T | undefined | void>): AsyncCollection<T> {
```

Reply:
Not sure about this.